File: src/token_saver/utils/token_counter.py
"""Approximate token counter utility.

Uses a character-based heuristic (~4 chars per token) for fast estimation.
Optionally uses tiktoken for exact OpenAI tokenizer counts.
"""

from __future__ import annotations


def estimate_tokens(text: str) -> int:
    """Estimate token count using character-based heuristic.

    Uses the industry-standard approximation of ~4 characters per token,
    which is accurate within ±10% for English text and code.
    """
    if not text:
        return 0
    return max(1, len(text) // 4)


def count_tokens_exact(text: str, model: str = "gpt-4o") -> int:
    """Count exact tokens using tiktoken (requires 'counting' extra).

    Falls back to heuristic estimation if tiktoken is not installed.
    """
    try:
        import tiktoken

        encoding = tiktoken.encoding_for_model(model)
        return len(encoding.encode(text))
    except (ImportError, KeyError):
        return estimate_tokens(text)


def format_savings(original: str, optimized: str) -> str:
    """Format a human-readable savings summary.

    Returns a string like: '847 → 127 tokens (85.0% saved)'
    """
    orig_tokens = estimate_tokens(original)
    opt_tokens = estimate_tokens(optimized)

    if orig_tokens == 0:
        return "0 → 0 tokens (no content)"

    saved_pct = (1 - opt_tokens / orig_tokens) * 100
    return f"{orig_tokens} → {opt_tokens} tokens ({saved_pct:.1f}% saved)"

File: crates/token-saver-core/src/token_counter.rs
//! Fast heuristic token estimation matching OpenAI / Claude approximations.

/// Estimate tokens for a given string slice (~4 chars per token).
pub fn estimate_tokens(text: &str) -> usize {
    if text.is_empty() {
        return 0;
    }
    let char_count = text.chars().count();
    (char_count / 4).max(1)
}

/// Format token savings summary string.
pub fn format_savings(original: &str, optimized: &str) -> String {
    let orig = estimate_tokens(original);
    let opt = estimate_tokens(optimized);
    let saved = orig.saturating_sub(opt);
    let pct = if orig > 0 {
        ((saved as f64 / orig as f64) * 100.0) as u32
    } else {
        0
    };
    format!("{orig} -> {opt} tokens ({pct}% saved)")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_estimate_tokens() {
        assert_eq!(estimate_tokens(""), 0);
        assert_eq!(estimate_tokens("hello"), 1);
        assert_eq!(estimate_tokens(&"a".repeat(400)), 100);
    }

    #[test]
    fn test_format_savings() {
        let orig = "a".repeat(400);
        let opt = "a".repeat(40);
        let res = format_savings(&orig, &opt);
        assert!(res.contains("100 -> 10"));
        assert!(res.contains("90% saved"));
    }
}
